File: riistaandroid/src/main/java/fi/riista/mobile/adapter/ShootingTestUserAttempsAdapter.java
package fi.riista.mobile.adapter;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.widget.AppCompatImageButton;

import java.util.List;

import fi.riista.mobile.R;
import fi.riista.mobile.RiistaApplication;
import fi.riista.mobile.activity.ShootingTestEditAttemptActivity;
import fi.riista.mobile.activity.ShootingTestUserAttemptsActivity;
import fi.riista.mobile.models.shootingTest.ShootingTestAttemptDetailed;
import fi.riista.mobile.network.shootingTest.DeleteAttemptForParticipantTask;

import static fi.riista.mobile.activity.ShootingTestEditAttemptActivity.EXTRA_ATTEMPT;
import static fi.riista.mobile.activity.ShootingTestEditAttemptActivity.EXTRA_ATTEMPT_ID;
import static fi.riista.mobile.activity.ShootingTestEditAttemptActivity.EXTRA_PARTICIPANT_ID;
import static fi.riista.mobile.activity.ShootingTestEditAttemptActivity.EXTRA_PARTICIPANT_NAME;
import static fi.riista.mobile.activity.ShootingTestEditAttemptActivity.EXTRA_PARTICIPANT_REV;

public class ShootingTestUserAttempsAdapter extends ArrayAdapter<ShootingTestAttemptDetailed> {

    private ShootingTestUserAttemptsActivity mContext;

    private long mParticipantId;
    private int mParticipantRev;
    private String mParticipantName;
    private boolean canEdit;

    private View.OnClickListener mOnEditClickListener = new View.OnClickListener() {
        @Override
        public void onClick(final View v) {
            final int position = (int) v.getTag();
            final ShootingTestAttemptDetailed clickedItem = getItem(position);

            if (clickedItem != null) {
                final Intent intent = new Intent(mContext, ShootingTestEditAttemptActivity.class);
                intent.putExtra(EXTRA_ATTEMPT, clickedItem);
                intent.putExtra(EXTRA_ATTEMPT_ID, clickedItem.id);
                intent.putExtra(EXTRA_PARTICIPANT_ID, mParticipantId);
                intent.putExtra(EXTRA_PARTICIPANT_REV, mParticipantRev);
                intent.putExtra(EXTRA_PARTICIPANT_NAME, mParticipantName);

                mContext.setEditTypeLimitsTo(intent, clickedItem.type);

                mContext.startActivity(intent);
            }
        }
    };
    private View.OnClickListener mOnDeleteClickListener = new View.OnClickListener() {
        @Override
        public void onClick(final View v) {
            final int position = (int) v.getTag();
            final ShootingTestAttemptDetailed clickedItem = getItem(position);

            if (clickedItem != null) {
                new AlertDialog.Builder(mContext)
                        .setMessage(mContext.getString(R.string.confirm_delete_attempt))
                        .setPositiveButton(R.string.yes, (dialog, which) -> deleteAttempt(clickedItem.id))
                        .setNegativeButton(R.string.no, null)
                        .show();
            }
        }
    };

    public ShootingTestUserAttempsAdapter(final List<ShootingTestAttemptDetailed> data,
                                          final long participantId,
                                          final boolean completed,
                                          @NonNull final ShootingTestUserAttemptsActivity context) {

        super(context, R.layout.view_shooting_test_attempt_item, data);

        mContext = context;
        mParticipantId = participantId;
        canEdit = !completed;
    }

    public void setParticipant(final long participantId,
                               final int participantRev,
                               final String participantName,
                               final boolean completed) {

        mParticipantId = participantId;
        mParticipantRev = participantRev;
        mParticipantName = participantName;
        canEdit = !completed;
    }

    private void deleteAttempt(final long attemptId) {
        final DeleteAttemptForParticipantTask task = new DeleteAttemptForParticipantTask(
                RiistaApplication.getInstance().getWorkContext(), attemptId) {

            @Override
            protected void onFinishText(final String text) {
                mContext.refreshData();
            }

            @Override
            protected void onError() {
                super.onError();

                final String errorMsg =
                        String.format(mContext.getString(R.string.error_operation_failed), getHttpStatusCode());

                Toast.makeText(mContext, errorMsg, Toast.LENGTH_LONG).show();
            }
        };
        task.start();
    }

    @NonNull
    @Override
    public View getView(final int position, View convertView, @NonNull final ViewGroup parent) {
        final Context context = getContext();
        final ViewHolder viewHolder;

        if (convertView == null) {
            final LayoutInflater inflater = LayoutInflater.from(context);
            convertView = inflater.inflate(R.layout.view_shooting_test_attempt_item, parent, false);

            viewHolder = new ViewHolder();
            viewHolder.attemptType = convertView.findViewById(R.id.attempt_type);
            viewHolder.result = convertView.findViewById(R.id.attempt_result);
            viewHolder.hits = convertView.findViewById(R.id.attempt_hits);

            viewHolder.editButton = convertView.findViewById(R.id.edit_button);
            viewHolder.deleteButton = convertView.findViewById(R.id.delete_button);

            viewHolder.editButton.setOnClickListener(mOnEditClickListener);
            viewHolder.deleteButton.setOnClickListener(mOnDeleteClickListener);

            convertView.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) convertView.getTag();
        }

        final ShootingTestAttemptDetailed item = getItem(position);

        if (item != null) {
            viewHolder.attemptType.setText(ShootingTestAttemptDetailed.localizedTypeText(context, item.type));
            viewHolder.result.setText(ShootingTestAttemptDetailed.localisedResultText(context, item.result));
            viewHolder.hits.setText(String.valueOf(item.hits));

            viewHolder.editButton.setEnabled(canEdit);
            viewHolder.editButton.setTag(position);

            viewHolder.deleteButton.setEnabled(canEdit);
            viewHolder.deleteButton.setTag(position);
        }

        return convertView;
    }

    private static class ViewHolder {
        TextView attemptType;
        TextView result;
        TextView hits;
        AppCompatImageButton editButton;
        AppCompatImageButton deleteButton;
    }
}

File: riistaandroid/src/main/java/fi/riista/mobile/adapter/ShootingTestUserAttempsAdapter.kt
package fi.riista.mobile.adapter

import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import androidx.appcompat.widget.AppCompatImageButton
import fi.riista.mobile.R
import fi.riista.mobile.activity.ShootingTestEditAttemptActivity
import fi.riista.mobile.activity.ShootingTestUserAttemptsActivity
import fi.riista.mobile.models.shootingTest.ShootingTestAttemptDetailed
import fi.riista.mobile.models.shootingTest.ShootingTestAttemptDetailed.Companion.localisedResultText
import fi.riista.mobile.models.shootingTest.ShootingTestAttemptDetailed.Companion.localizedTypeText

class ShootingTestUserAttempsAdapter(
    data: List<ShootingTestAttemptDetailed?>?,
    private var participantId: Long,
    completed: Boolean,
    private val context: ShootingTestUserAttemptsActivity,
    private val deleteClickedListener: (attemptId: Long) -> Unit,
) : ArrayAdapter<ShootingTestAttemptDetailed?>(
    context, R.layout.view_shooting_test_attempt_item, data!!
) {
    private var participantRev = 0
    private var participantName: String? = null
    private var canEdit: Boolean
    private val onEditClickListener = View.OnClickListener { v ->
        val position = v.tag as Int
        val clickedItem = getItem(position)
        if (clickedItem != null) {
            val intent = Intent(context, ShootingTestEditAttemptActivity::class.java)
            intent.putExtra(ShootingTestEditAttemptActivity.EXTRA_ATTEMPT, clickedItem)
            intent.putExtra(ShootingTestEditAttemptActivity.EXTRA_ATTEMPT_ID, clickedItem.id)
            intent.putExtra(ShootingTestEditAttemptActivity.EXTRA_PARTICIPANT_ID, participantId)
            intent.putExtra(ShootingTestEditAttemptActivity.EXTRA_PARTICIPANT_REV, participantRev)
            intent.putExtra(ShootingTestEditAttemptActivity.EXTRA_PARTICIPANT_NAME, participantName)
            context.setEditTypeLimitsTo(intent, clickedItem.type)
            context.startActivity(intent)
        }
    }
    private val onDeleteClickListener = View.OnClickListener { v ->
        val position = v.tag as Int
        val clickedItem = getItem(position)
        if (clickedItem != null) {
            deleteClickedListener(clickedItem.id)
        }
    }

    init {
        canEdit = !completed
    }

    fun setParticipant(
        participantId: Long,
        participantRev: Int,
        participantName: String?,
        completed: Boolean
    ) {
        this.participantId = participantId
        this.participantRev = participantRev
        this.participantName = participantName
        canEdit = !completed
    }

    override fun getView(position: Int, view: View?, parent: ViewGroup): View {
        var convertView = view
        val context = context
        val viewHolder: ViewHolder
        if (convertView == null) {
            val inflater = LayoutInflater.from(context)
            convertView = inflater.inflate(R.layout.view_shooting_test_attempt_item, parent, false)
            viewHolder = ViewHolder()
            viewHolder.attemptType = convertView.findViewById(R.id.attempt_type)
            viewHolder.result = convertView.findViewById(R.id.attempt_result)
            viewHolder.hits = convertView.findViewById(R.id.attempt_hits)
            viewHolder.editButton = convertView.findViewById(R.id.edit_button)
            viewHolder.deleteButton = convertView.findViewById(R.id.delete_button)
            viewHolder.editButton.setOnClickListener(onEditClickListener)
            viewHolder.deleteButton.setOnClickListener(onDeleteClickListener)
            convertView.tag = viewHolder
        } else {
            viewHolder = convertView.tag as ViewHolder
        }
        val item = getItem(position)
        if (item != null) {
            viewHolder.attemptType.text = localizedTypeText(context, item.type)
            viewHolder.result.text = localisedResultText(context, item.result)
            viewHolder.hits.text = item.hits.toString()
            viewHolder.editButton.isEnabled = canEdit
            viewHolder.editButton.tag = position
            viewHolder.deleteButton.isEnabled = canEdit
            viewHolder.deleteButton.tag = position
        }
        return convertView!!
    }

    private class ViewHolder {
        lateinit var attemptType: TextView
        lateinit var result: TextView
        lateinit var hits: TextView
        lateinit var editButton: AppCompatImageButton
        lateinit var deleteButton: AppCompatImageButton
    }
}
